lib/ProductOpener/Products.pm

```diff
@@ -2955,6 +2955,83 @@ sub compute_languages ($product_ref) {
 	return;
 }
 
+=head2 review_product_type ( $product_ref )
+
+Reviews the product type based on the presence of specific tags in the categories field.
+Updates the product type if necessary.
+
+=head3 Arguments
+
+=head4 Product reference $product_ref
+
+A reference to a hash containing the product details.
+
+=cut
+
+sub review_product_type ($product_ref) {
+
+	my $error;
+
+	if ($product_ref->{product_type} eq "beauty") {
+		if (has_tag($product_ref, "categories", "en:non-beauty-products")) {
+			if (has_tag($product_ref, "categories", "en:open-food-facts")) {
+				$error = change_product_type($product_ref, "food");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-pet-food-facts")) {
+				$error = change_product_type($product_ref, "petfood");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-products-facts")) {
+				$error = change_product_type($product_ref, "product");
+			}
+		}
+	}
+	elsif ($product_ref->{product_type} eq "food") {
+		if (has_tag($product_ref, "categories", "en:non-food-products")) {
+			if (has_tag($product_ref, "categories", "en:open-beauty-facts")) {
+				$error = change_product_type($product_ref, "beauty");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-pet-food-facts")) {
+				$error = change_product_type($product_ref, "petfood");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-products-facts")) {
+				$error = change_product_type($product_ref, "product");
+			}
+		}
+	}
+	elsif ($product_ref->{product_type} eq "petfood") {
+		if (has_tag($product_ref, "categories", "en:non-pet-food-products")) {
+			if (has_tag($product_ref, "categories", "en:open-beauty-facts")) {
+				$error = change_product_type($product_ref, "beauty");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-food-facts")) {
+				$error = change_product_type($product_ref, "food");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-products-facts")) {
+				$error = change_product_type($product_ref, "product");
+			}
+		}
+	}
+	elsif ($product_ref->{product_type} eq "product") {
+		if (has_tag($product_ref, "categories", "en:non-food-products")) {
+			if (has_tag($product_ref, "categories", "en:open-beauty-facts")) {
+				$error = change_product_type($product_ref, "beauty");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-food-facts")) {
+				$error = change_product_type($product_ref, "food");
+			}
+			elsif (has_tag($product_ref, "categories", "en:open-pet-food-facts")) {
+				$error = change_product_type($product_ref, "petfood");
+			}
+		}
+	}
```

Comment:
Thanks @benbenben2 that's great.

That said, I would have first computed "expected_product_type" using "has_tag" and if it's defined then, I would have compared it to $product_ref->{product_type}.

It would allow less repetitions in the code. 

Are you willing to refactor ?

Reply:
Sure. I applied some changes. Is it what you expected?

Reply:
it was more:

```perl
	my $expected_type;
	if (has_tag($product_ref, "categories", "en:open-food-facts")) {
		$expected_type = "food";
	}
	elsif (has_tag($product_ref, "categories", "en:open-pet-food-facts")) {
		$expected_type = "petfood";
	}
	elsif (has_tag($product_ref, "categories", "en:open-products-facts")) {
		$expected_type = "product";
	}
	elsif (has_tag($product_ref, "categories", "en:open-beauty-facts")) {
		$expected_type = "beauty";
	}
	if ($expected_type and ($product_ref->{product_type} ne expected_type)) {
		$error = change_product_type($product_ref, expected_type);
	}
```